### PR TITLE
[jsk_perception/apply_mask_image] Add negative option

### PR DESCRIPTION
**Pull request by iory**
original image
image_publisher

mask image
mask_publisher

apply_mask image
apply_mask_image

apply_negative_mask 
apply_negative_mask_image

### CODE REVIEW

**Comment by wkentaro** [comment]
how to handle clip with negative?

**Comment by iory** [comment]
in https://github.com/iory/jsk_recognition/blob/f579faff74d5b68fb529fdf8823dd2b37b184b34/jsk_perception/src/apply_mask_image.cpp#L110-L119,
```
    if (negative_) {
      cv::bitwise_not(mask, mask);
    }

    if (clip_) {
      cv::Rect region = jsk_recognition_utils::boundingRectOfMaskImage(mask);
      mask = mask(region);
      image = image(region);
    }
```
Since processing to make it negative is before clip, there is no problem.
This option is intended to absorb the differences between the cases as follows.
For example, a person who wishes to retrieve a part of the robot can accomplish by applying apply_mask_image to robot_self_filter, but for a person wanting to extract the part of the robot, that negative is necessary.
mask_publisher
image_publisher
apply_mask_image
apply_negative_mask_image

**Comment by wkentaro** [comment]
I see. But usually if `negative=True` and `clip=True` ends up no clipping because clip finds the bounding box of white region.
I'm not sure this is okay.

**Comment by iory** [comment]
OK, add nagative_before_clip option. 
this solves order of clipping and negative

File: jsk_perception/src/apply_mask_image.cpp [hunk 1]
```diff
@@ -105,11 +107,25 @@ namespace jsk_perception
...
+    } else {
+      if (clip_) {
+        cv::Rect region = jsk_recognition_utils::boundingRectOfMaskImage(mask);
+        mask = mask(region);
+        image = image(region);
+      }
+      if (negative_) {
+        cv::bitwise_not(mask, mask);
+      }
     }
```
Comment (wkentaro) [inline jsk_perception/src/apply_mask_image.cpp]: Why not below?

```bash
if (negative_before_clip_ && negative_) {
  cv::bitwise_not(...)
}

if (clip_) {
  # clip
}

if (!negative_before_clip_ && negative_) {
  cv::bitwise_not(...)
}
```

Comment (iory) [inline jsk_perception/src/apply_mask_image.cpp]: Changed

**Comment by wkentaro** [comment]
LGTM

**Comment by wkentaro** [comment]
Ah, wait. Maybe you should add some descriptions about the new parameters in doc @iory.

**Comment by iory** [comment]
OK, I add

**Comment by wkentaro** [comment]
LGTM

**Pull request merged**